### DISCUSSION TITLE
Is it possible to trace another eBPF program using fentry/fexit hooks and bpftrace?

### DISCUSSION THREAD

**Discussion by mscastanho** [bpftrace/bpftrace#3507] [Q&A] [answered]
I'd like to trace execution times of a eBPF program attached to XDP hook using bpftrace. Basically attaching to fentry and fexit hooks to get timestamps and calculate runtime. Is it possible?

This is what I tried: I looked for my program's ksym name in `/proc/kallsyms` and passed that to bpftrace, but got a BTF-related error:
```
# /tmp/bpftrace -e 'kfunc:bpf_prog_35bb045a2eb16e2d_filter  { @start[tid] = nsecs; } kretfunc:bpf_prog_35bb045a2eb16e2d_filter  /@start[tid]/ { @ns = lhist(nsecs - @start[tid], 256, 8000, 100); }'
Attaching 2 probes...
ERROR: No BTF found for bpf_prog_35bb045a2eb16e2d_filter
```

I imagine `bpftrace` looks for BTF data in `/sys/kernel/btf/vmlinux`, but since my program is not part of the default vmlinux it can't find BTF data related to it.

**Comment by danobi** [comment, marked as answer]
I suspect bpftrace cannot do exactly what you're asking for b/c IIUC fentry/fexit probes require BTF information. And I think kprobe doesn't work either b/c the prog is JITed and thus not present in `/sys/kernel/debug/tracing/available_filter_functions`.

You might be able to get most of what you want thru:

```
       bpftool prog { show | list } [PROG]
              <snip>

              Since  Linux 5.1 the kernel can collect statistics on BPF programs (such as the total time spent running the program, and the number of times
              it was run). If available, bpftool shows such statistics. However, the kernel does not collect them by defaults, as it slightly impacts  per‐
              formance on each program run. Activation or deactivation of the feature is performed via the kernel.bpf_stats_enabled sysctl knob.
```

**Reply by mscastanho** [reply]
True. But it would also be nice to also use other bpftrace features, like histograms.
But thanks for the answer =)
